examples/chat/src/main.rs

```diff
@@ -1,37 +1,88 @@
 mod chat;
-mod room;
+mod models;
+mod routes;
+mod schema;
+mod util;
 
-use http::header;
 use std::process::ExitCode;
-use via::{App, Error, Response, Server};
+use via::error::{Error, Rescue};
+use via::{App, Cookies, Server, ws};
 
-use crate::chat::Chat;
+use chat::Chat;
+use util::Auth;
 
-const CSP: &str = "default-src 'self'; connect-src 'self'";
+const SESSION: &str = "via-chat-session";
+
+type Request = via::Request<Chat>;
+type Next = via::Next<Chat>;
 
 #[tokio::main]
 async fn main() -> Result<ExitCode, Error> {
-    let mut app = App::new(Chat::new());
+    dotenvy::dotenv()?;
+
+    let mut app = {
+        let pool = chat::establish_pg_connection().await;
+        let secret = chat::load_session_secret();
+
+        App::new(Chat::new(pool, secret))
+    };
+
+    app.uses(Cookies::new().allow(SESSION));
+    app.uses(Auth::new(SESSION));
+
+    app.route("/").to(via::get(routes::home));
+
+    let mut api = app.route("/api");
+
+    api.uses(Rescue::with(util::error_sanitizer));
+
+    // Perform a websocket upgrade and start chatting.
+    api.route("/subscribe").to(ws::upgrade(routes::subscribe));
+
+    api.route("/threads").scope(|threads| {
+        let mut thread = {
+            let (collection, member) = via::rest!(routes::threads);
+
+            threads.route("/").to(collection);
+            threads.route("/:thread-id").to(member)
+        };
+
+        // 403 when the current user is not in the thread.
+        thread.uses(routes::threads::authorization);
```

Comment:
We'll probably attempt to statically determine if a user is participating in a thread. Perhaps we'll have a list of threads to which they belong in the cookie or JWT. This doesn't scale to an unbounded number of threads but it's probably fine for this example. This is pretty robust as it is.

It's possible that we'll defer authorization until the database query is made and confirm that a join table exists with the user_id and thread_id as a part of the database query and use the `FoundOrForbidden` trait to propagate authorization errors like we do for messages and reactions.